service/create/asg_stack.go

```diff
@@ -246,16 +257,55 @@ func (s *Service) updateASGStack(input asgStackInput) error {
 		Client: input.clients.CloudFormation,
 
 		// Settings.
-		ASGMaxSize:  input.asgSize,
-		ASGMinSize:  input.asgSize,
-		ImageID:     imageID,
-		Name:        key.AutoScalingGroupName(input.cluster, input.asgType),
-		TemplateURL: templateURL,
+		ASGMaxSize:            input.asgSize,
+		ASGMinSize:            input.asgSize,
+		ImageID:               imageID,
+		MaxBatchSize:          getMaxBatchSize(input.asgSize),
+		MinInstancesInService: getMinInstancesInService(input.asgSize),
+		Name: key.AutoScalingGroupName(input.cluster, input.asgType),
+		RollingUpdatePauseTime: rollingUpdatePauseTime,
+		TemplateURL:            templateURL,
+	}
+
+	stackUpdated, err := stack.Update()
+	if err != nil {
+		return stackUpdated, microerror.Mask(err)
+	}
+
+	return stackUpdated, nil
+}
+
+// getMaxBatchSize calculates the max batch size for the rolling update policy.
+func getMaxBatchSize(asgSize int) int {
+	var batchSize int
+
+	switch {
+	case asgSize <= 2:
+		batchSize = 1
+	case asgSize <= 4:
+		batchSize = 2
+	default:
+		// Calculate batch size and round up to nearest int.
+		result := float64(asgSize) * asgMaxBatchSizeRatio
+		batchSize = int(result + 0.5)
 	}
 
-	if err := stack.Update(); err != nil {
-		return microerror.Mask(err)
+	return batchSize
+}
+
+// getMinInstancesInService calculates the min number for instances to keep in
+// service for the rolling update policy.
+func getMinInstancesInService(asgSize int) int {
+	var minInstances int
+
+	switch {
+	case asgSize <= 2:
+		minInstances = 1
+	default:
+		// Calculate min instances and round up to nearest int.
+		result := float64(asgSize) * asgMinInstancesRatio
+		minInstances = int(result + 0.5)
```

Comment:
- To confirm the operator isn't changing the TPO / CRO. The CoreOS AMI needs to be updated externally in the TPO. I agree an operator shouldn't update its own TPO.

- From the updates meeting, I thought we were going to add an endpoint to kubernetesd? We would use this internally to trigger the update. But this could also be done by an upgrades controller.

- On node IDs yes I think they are needed for deleting / updating specific nodes. The EC2  instances have their own internal IDs so we need to tag the nodes with our node IDs. This is possible via the EC2 API directly but not via the ASG because its tags apply to all nodes.

- On whether to use ASGs we could launch EC2 instances manually and do a reconciliation. We could then use the node IDs and the operator would be more like the KVM operator. The downside is we have to implement the ASG functionality ourselves.

- An example is the upcoming change to spread nodes across AZs. With an ASG we just specify the AZs and it does the balancing. Launching instances manually we'd need to do that ourselves. This is why I'm still in favour of using ASGs.